File: python/examples/mcp_proxy_demo.py
"""The agent on the tool path: aegis-mcp-proxy minimizes tool output before the model sees it.

The primary target is the coding agent itself (Claude Code, Cursor, codex, agy,
any MCP host). You do not change the agent or the MCP server — you point the
agent's MCP config at `aegis-mcp-proxy` instead of the raw server:

    host  <--stdio-->  aegis-mcp-proxy --policy p.json  <--stdio-->  real MCP server

Every tools/call *result* is filtered to the policy's scope before it can enter
the agent's model context, logs, or downstream tool calls — server unmodified,
fail-closed on tools the policy does not map.

This script plays the "host". It runs the SAME tool call twice — once straight
against the server, once through the proxy — and prints what the agent receives
each time. No API key, no network, no Aegis Core.

    pip install aegis-trust
    python examples/mcp_proxy_demo.py
"""

import json
import subprocess
import sys
import tempfile
from pathlib import Path

# A stand-in MCP server. Its tool returns an over-broad customer record — the
# kind of tool output that, unfiltered, lands whole in your agent's context.
FAKE_SERVER = r"""
import json, sys
RECORD = {"name": "ACME", "company": "ACME Inc",
          "email": "ceo@acme.test", "ssn": "123-45-6789",
          "card_number": "4242-4242-4242-4242"}
for line in sys.stdin:
    line = line.strip()
    if not line:
        continue
    msg = json.loads(line)
    if "id" not in msg:
        continue
    if msg.get("method") == "tools/call":
        result = {"structuredContent": RECORD,
                  "content": [{"type": "text", "text": json.dumps(RECORD)}]}
    else:
        result = {"ok": True}
    sys.stdout.write(json.dumps({"jsonrpc": "2.0", "id": msg["id"], "result": result}) + "\n")
    sys.stdout.flush()
"""

# The policy: for this tool, the agent's purpose only needs name + company.
POLICY = {
    "policy_schema_version": 0,
    "policy_id": "demo",
    "purposes": {"customer_data": {"scope": ["name", "company"]}},
    "tools": {"query_business_data": "customer_data"},
    "defaults": {"unknown_tool": "block"},
}

CALL = {
    "jsonrpc": "2.0",
    "id": 1,
    "method": "tools/call",
    "params": {"name": "query_business_data", "arguments": {}},
}
BAD_CALL = {
    "jsonrpc": "2.0",
    "id": 2,
    "method": "tools/call",
    "params": {"name": "exfiltrate_everything", "arguments": {}},
}


def rpc(proc, msg):
    proc.stdin.write(json.dumps(msg) + "\n")
    proc.stdin.flush()
    return json.loads(proc.stdout.readline())


def main() -> int:
    with tempfile.TemporaryDirectory() as d:
        d = Path(d)
        server = d / "server.py"
        server.write_text(FAKE_SERVER)
        policy = d / "policy.json"
        policy.write_text(json.dumps(POLICY))
        audit = d / "events.jsonl"

        # 1) WITHOUT the proxy: the agent talks straight to the server.
        raw = subprocess.Popen(
            [sys.executable, str(server)],
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            text=True,
        )
        got = rpc(raw, CALL)["result"]["structuredContent"]
        raw.stdin.close()
        raw.wait(timeout=10)
        print("WITHOUT proxy — what the agent's model receives:")
        print("  ", got)
        print("   -> ssn/card in agent context:", "ssn" in got or "card_number" in got)

        # 2) WITH aegis-mcp-proxy on the path (server unmodified).
        proxy = subprocess.Popen(
            [
                sys.executable,
                "-m",
                "aegis_trust.mcp_proxy",
                "--policy",
                str(policy),
                "--agent-id",
                "demo-host",
                "--audit-path",
                str(audit),
                "--",
                sys.executable,
                str(server),
            ],
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
        )
        through = rpc(proxy, CALL)["result"]["structuredContent"]
        blocked = rpc(proxy, BAD_CALL)["result"]
        proxy.stdin.close()
        proxy.wait(timeout=10)

        print("\nWITH aegis-mcp-proxy — what the agent's model receives:")
        print("  ", through)
        print(
            "   -> ssn/card in agent context:",
            "ssn" in through or "card_number" in through,
            "(stripped at the process boundary)",
        )

        print("\nA tool the policy does not allow never even runs:")
        print(
            "   exfiltrate_everything ->",
            "BLOCKED" if blocked.get("isError") else "forwarded",
            "-",
            blocked["content"][0]["text"],
        )

        # The boundary leaves an audit trail (enforcement_point=mcp_proxy).
        events = [json.loads(x) for x in audit.read_text().splitlines() if x.strip()]
        kinds = [
            (e.get("enforcement_point"), e.get("decision") or e.get("outcome"))
            for e in events
        ]
        print("\nCanonical audit events emitted:", kinds)

    print(
        "\nTakeaway: the agent is on the tool path through the proxy. It cannot "
        "see fields outside the declared scope, and cannot call a tool the "
        "policy does not map — one line of MCP config, server unmodified."
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: node/examples/mcpProxyDemo.ts
// The agent on the tool path: aegis-mcp-proxy minimizes tool output before the
// model sees it. Node mirror of python/examples/mcp_proxy_demo.py.
//
// Point any MCP host (Claude Code, Cursor, codex, agy) at `aegis-mcp-proxy`
// instead of the raw server and every tools/call result is filtered to policy
// before it can enter the agent's model context. This script plays the host and
// runs the same call twice — straight against the server, then through the proxy.
//
//   npm install aegis-trust && npm run build
//   npx tsx examples/mcpProxyDemo.ts

import { spawn } from "node:child_process";
import { mkdtempSync, readFileSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { fileURLToPath } from "node:url";

const PROXY = fileURLToPath(new URL("../dist/mcpProxy.js", import.meta.url));

// A stand-in MCP server whose tool returns an over-broad customer record.
const FAKE_SERVER = `
import { createInterface } from "node:readline";
const RECORD = { name: "ACME", company: "ACME Inc", email: "ceo@acme.test",
  ssn: "123-45-6789", card_number: "4242-4242-4242-4242" };
createInterface({ input: process.stdin }).on("line", (line) => {
  if (!line.trim()) return;
  const msg = JSON.parse(line);
  if (!("id" in msg)) return;
  const result = msg.method === "tools/call"
    ? { structuredContent: RECORD }
    : { ok: true };
  process.stdout.write(JSON.stringify({ jsonrpc: "2.0", id: msg.id, result }) + "\\n");
});
`;

const POLICY = {
  policy_schema_version: 0,
  policy_id: "demo",
  purposes: { customer_data: { scope: ["name", "company"] } },
  tools: { query_business_data: "customer_data" },
  defaults: { unknown_tool: "block" },
};

const CALL = { jsonrpc: "2.0", id: 1, method: "tools/call", params: { name: "query_business_data", arguments: {} } };
const BAD = { jsonrpc: "2.0", id: 2, method: "tools/call", params: { name: "exfiltrate_everything", arguments: {} } };

function collect(args: string[], input: string): Promise<{ out: string; auditPath?: string }> {
  return new Promise((resolve) => {
    const proc = spawn(process.execPath, args, { stdio: ["pipe", "pipe", "inherit"] });
    let out = "";
    proc.stdout.on("data", (d) => (out += d.toString()));
    proc.stdin.write(input);
    proc.stdin.end();
    proc.on("exit", () => resolve({ out }));
  });
}

function byId(out: string): Map<number, Record<string, unknown>> {
  const m = new Map<number, Record<string, unknown>>();
  for (const line of out.split("\n")) {
    if (!line.trim()) continue;
    const o = JSON.parse(line);
    m.set(o.id, o);
  }
  return m;
}

async function main(): Promise<void> {
  const dir = mkdtempSync(join(tmpdir(), "aegis-demo-"));
  const server = join(dir, "server.mjs");
  const policy = join(dir, "policy.json");
  const audit = join(dir, "events.jsonl");
  writeFileSync(server, FAKE_SERVER);
  writeFileSync(policy, JSON.stringify(POLICY));

  // 1) WITHOUT the proxy: the agent talks straight to the server.
  const raw = await collect([server], JSON.stringify(CALL) + "\n");
  const got = byId(raw.out).get(1)!.result as Record<string, unknown>;
  const gotData = (got.structuredContent ?? got) as Record<string, unknown>;
  console.log("WITHOUT proxy — what the agent's model receives:");
  console.log("  ", gotData);
  console.log("   -> ssn/card in agent context:", "ssn" in gotData || "card_number" in gotData);

  // 2) WITH aegis-mcp-proxy on the path (server unmodified).
  const through = await collect(
    [PROXY, "--policy", policy, "--agent-id", "demo-host", "--audit-path", audit, "--", process.execPath, server],
    JSON.stringify(CALL) + "\n" + JSON.stringify(BAD) + "\n",
  );
  const m = byId(through.out);
  const thru = (m.get(1)!.result as Record<string, unknown>).structuredContent as Record<string, unknown>;
  console.log("\nWITH aegis-mcp-proxy — what the agent's model receives:");
  console.log("  ", thru);
  console.log("   -> ssn/card in agent context:", "ssn" in thru || "card_number" in thru, "(stripped at the process boundary)");

  const blocked = m.get(2)!.result as { isError?: boolean; content: Array<{ text: string }> };
  console.log("\nA tool the policy does not allow never even runs:");
  console.log("   exfiltrate_everything ->", blocked.isError ? "BLOCKED" : "forwarded", "-", blocked.content[0].text);

  const events = readFileSync(audit, "utf8").trim().split("\n").filter(Boolean).map((l) => JSON.parse(l));
  console.log("\nCanonical audit events emitted:", events.map((e) => [e.enforcement_point, e.decision]));

  console.log(
    "\nTakeaway: the agent is on the tool path through the proxy. It cannot see " +
      "fields outside the declared scope, and cannot call a tool the policy does " +
      "not map — one line of MCP config, server unmodified.",
  );
}

main();
